### PR TITLE
New math define `ADDCLAMP()` - changes some codes to use this instead of clamp()

### PR DESCRIPTION
**Pull request by EvilDragonfiend**
## About The Pull Request
Adds a new proc `ADDCLAMP(val_respected, val_to_add, low, high)`
the proc `clamp()` has an issue that it doesn't respect its original value
For example...
* let's say you have 200 hallucination.
* you got a drug that increases +5 hallucination
  * used syntax is `clamp(mob.hallu + 5, 0, 60)` - note that it uses `min(max(bla bla))` but it's identical to clamp()
* then somehow, the drug is supposed to increase your hallucination above +60, but it decreases the val to 60

Now, back to ADDCLAMP() proc...
the proc respects the original value in its special parameter `val_respected`

* back to the example... you have 200 hallucination.
* `hallucination = clamp(hallucination + 5, 0, 60)` will make 200 to 60.
* `hallucination = ADDCLAMP(hallucination, 5, 0, 60)` will keep 200 as is.

### Interesting calculation
image

We can use this to make the effect less effective when it reaches the cap.
A sample syntax `ADDCLAMP(start + 3, 6, 0, 30)` can mean:
* Add value 3 and 6
* but once it reaches the cap, only add the leftover
* but still, add 3 beyond the cap.

## Why It's Good For The Game
this is likely bug fix, making things corrected.

## Testing Photographs and Procedure

image
```DM
val 15, add 5,  min 0, max 10 = 15	(val 15 is prioritised. won't return 15+5)
val 5,  add 10, min 0, max 10 = 10	(max 10 is prioritised. won't return 5+10)
val 5,  add -10, min 0, max 10 = 0 	(min 0 is prioritised. won't return 5-10)
val -5, add -10, min 0, max 10 = -5	(val -5 is prioritised. won't return -5-10)
```

image

You can see it can return 80

## Changelog
:cl: EvilDragonfiend (Author), JixS4v (Def name suggestion)
code: ADDCLAMP() math define.
code: some codes that used clamp() now uses ADDCLAMP()
tweak: insanity anomaly now has max hallucination cap 100. It won't make you more insane than 100 points.
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
clamp_respected() is an incredibly confusing name, there is no easy explanation as to how it is being respected.
It should be something like add_until(), add_if_between() or add_between()

**Comment by EvilDragonfiend** [comment]
> clamp_respected() is an incredibly confusing name, there is no easy explanation as to how it is being respected. It should be something like add_until(), add_if_between() or add_between()

sadly the suggested names are not working as well. "add" part works, but "until" "between" can't express if the proc can return out of the range.

**Comment by PowerfulBacon** [comment]
Yea, cause you are adding if the number is between the range. If the number isn't between the range then it won't add and the original number will stay outside the range

File: code/game/objects/effects/anomalies/anomaly_hallucination.dm [hunk 1]
```diff
@@ -37,8 +37,7 @@
 
 		// Everyone else
 		var/dist = sqrt(1 / max(1, get_dist(near, location)))
-		near.hallucination += strength * dist
-		near.hallucination = clamp(near.hallucination, 0, 150)
+		near.hallucination = ADDCLAMP(near.hallucination, strength * dist, 0, 150)
```
Comment (EvilDragonfiend) [inline code/game/objects/effects/anomalies/anomaly_hallucination.dm:40]: This case is why I wanted to keep it as proc. If a given syntax is complex, you'll do that multiple times - which can be even worse than proc version.
If it's like `ADDCLAMP(something, nice_proc(), 0, 150)`, and when the nice_proc() takes 1 seconds to calculate, the worse scenario is checking it 3 times to return the last condition, 3 seconds.
At least this is a simple multiply syntax, and won't cause the issue I am worrying here, but you'll see what I see from here.

or declaring a temp var will be encouraged, but I am not sure if a future coder will know that.

File: code/__DEFINES/maths.dm [hunk 2]
```diff
@@ -239,3 +239,20 @@
 /// Gives the number of pixels in an orthogonal line of tiles.
 #define TILES_TO_PIXELS(tiles)			(tiles * PIXELS)
 // )
+
+/// Similar to clamp(). This is identical to 'x + y' thing, but add to 'x' will be limited within low-high range.
+/// If val_respected is already outside of low-high range, it will return val_respected, instead of the value within the range.
+#define ADDCLAMP(val_respected, val_to_add, low, high) ( \
+		((val_respected) + (val_to_add)) >= (high) \
```
Comment (JixS4v) [inline code/__DEFINES/maths.dm]:
```suggestion
		((val_respected) + (val_to_add)) > (high) \
```

File: code/__DEFINES/maths.dm [hunk 3]
```diff
@@ -239,3 +239,20 @@
...
+		? max((val_respected), (high)) \
+		: ((val_respected) + (val_to_add)) <= (low) \
```
Comment (JixS4v) [inline code/__DEFINES/maths.dm]:
```suggestion
		: ((val_respected) + (val_to_add)) < (low) \
```

**Review by JixS4v — CHANGES_REQUESTED** [review]
These will make it so that the ends of the clamp are excluded, which makes for unpredictable behaviour. EDIT: probably has the same outcome, but in this case you save a max()/min() call

File: code/game/objects/effects/anomalies/anomaly_hallucination.dm (hunk 1 above)
Comment (JixS4v) [reply code/game/objects/effects/anomalies/anomaly_hallucination.dm:40]: can you declare a temp var inside the define?

Comment (JixS4v) [reply code/game/objects/effects/anomalies/anomaly_hallucination.dm:40]: also, a sum is probably a much cheaper operation than a proc call, even if done three times.

Comment (EvilDragonfiend) [reply code/game/objects/effects/anomalies/anomaly_hallucination.dm:40]: > can you declare a temp var inside the define?

You can't. Unless DM supports an arrow function. It's just not possible logic afaik.

Comment (EvilDragonfiend) [reply code/game/objects/effects/anomalies/anomaly_hallucination.dm:40]: > also, a sum is probably a much cheaper operation than a proc call, even if done three times.

I don't get why you say sum here.

**Comment by JixS4v** [comment]
> > These will make it so that the ends of the clamp are excluded, which makes for unpredictable behaviour. EDIT: probably has the same outcome, but in this case you save a max()/min() call
> 
> That does early return, so it has a good point to have 'equal or more than' condition.
> I will leave a comment why it has equal

In this case an early return wouldn't be optimal since it results in a needless proc call

**Comment by EvilDragonfiend** [comment]
> > > These will make it so that the ends of the clamp are excluded, which makes for unpredictable behaviour. EDIT: probably has the same outcome, but in this case you save a max()/min() call
> > 
> > 
> > That does early return, so it has a good point to have 'equal or more than' condition.
> > I will leave a comment why it has equal
> 
> In this case an early return wouldn't be optimal since it results in a needless proc call

?? how

**Comment by JixS4v** [comment]
> > > > These will make it so that the ends of the clamp are excluded, which makes for unpredictable behaviour. EDIT: probably has the same outcome, but in this case you save a max()/min() call
> > > 
> > > 
> > > That does early return, so it has a good point to have 'equal or more than' condition.
> > > I will leave a comment why it has equal
> > 
> > In this case an early return wouldn't be optimal since it results in a needless proc call
> 
> ?? how

You get a call of min() or max(), which are procs right?

**Comment by EvilDragonfiend** [comment]
> > > > > These will make it so that the ends of the clamp are excluded, which makes for unpredictable behaviour. EDIT: probably has the same outcome, but in this case you save a max()/min() call
> > > > 
> > > > 
> > > > That does early return, so it has a good point to have 'equal or more than' condition.
> > > > I will leave a comment why it has equal
> > > 
> > > 
> > > In this case an early return wouldn't be optimal since it results in a needless proc call
> > 
> > 
> > ?? how
> 
> You get a call of min() or max(), which are procs right?

That isn't what I call expensive. `((val_respected) + (val_to_add))` is possibly expensive.

**Comment by JixS4v** [comment]
An addition is not expensive by any means compared to a proc call

**Comment by EvilDragonfiend** [comment]
> An addition is not expensive by any means compared to a proc call

It's not about addition. it's about the parameters that can accept procs.
Imagine when a proc that takes 10 seconds is given to the first parameter. If the first check fails, you need to run the second check and it means extra +10 seconds.

File: code/game/objects/effects/anomalies/anomaly_hallucination.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/game/objects/effects/anomalies/anomaly_hallucination.dm:40]: > can you declare a temp var inside the define?

You can use global/static variables to temporarilly store values. See my PR on virtual multi-z

**Comment by EvilDragonfiend** [comment]
image

in a syntax `((var=value)|1)`, for the reason why it has `|1` is that because `(var=0)` is considered to be FALSE which doesn't execute other operators. Very funny trick.

**Comment by JixS4v** [comment]
You don't need to put low and high in glob variables since there's no optimization benefit given you're not optimizing away any operations. In fact, you should just premake the addition, this seems super wasteful.

File: code/__DEFINES/maths.dm [hunk 4]
```diff
@@ -239,3 +239,35 @@
...
+		((GLOB.addclamp_add = (val_to_add))||1) && \
+		((GLOB.addclamp_low = (low))||1) && \
+		((GLOB.addclamp_high = (high))||1) \
+	) && \
+	(GLOB.addclamp_respect + GLOB.addclamp_add) >= GLOB.addclamp_high \
+	? max(GLOB.addclamp_respect, GLOB.addclamp_high) \
+	: (GLOB.addclamp_respect + GLOB.addclamp_add) <= GLOB.addclamp_low \
+	? min(GLOB.addclamp_respect, GLOB.addclamp_low) \
+	: GLOB.addclamp_respect + GLOB.addclamp_add \
+)
```
Comment (PowerfulBacon) [inline code/__DEFINES/maths.dm:264]:
```
(val + add) > max ? (val + add) : val < min ? val : clamp(val + add, min, max)
```

val 15, add 5, low 0, high 15 (15 + 5 > 15, so returns 15).
val 5, add 10, low 0, high 10 (5 + 10 = 15, so returns 10)
val 5, add -10, low 0, high 10 (5 - 10 = -5, so returns 0)
val -5, add -10, low 0, high 10 (-5 - 15 = -15, so returns -5)

This passes the basic tests although I haven't fully understood the point of the edge cases so needs double checking

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

File: code/__DEFINES/maths.dm (hunk 4 above)
Comment (PowerfulBacon) [reply code/__DEFINES/maths.dm:264]: Wait why even make these global variables, they are litreally just assigned to values?

**Comment by EvilDragonfiend** [comment]
I will fix the stuff as what's been discussed in coderbus when I have time

**Comment by EvilDragonfiend** [comment]
I am convinved to do this as if, not add_clamp. I will repost a different version
